framework/calico_framework.py

```diff
@@ -0,0 +1,707 @@
+
+    def __repr__(self):
+        return "Offer(id=%s, tasks=%d, slave=%s, reserved_by=%s)" % (
+            self.offer_id[-5:], self.num_tasks_launchable,
+            self.offer.slave_id.value[-5:], self.reserved_by)
+
+
+class TestScheduler(mesos.interface.Scheduler):
+    def __init__(self, implicit_acknowledgements):
+        self.implicitAcknowledgements = implicit_acknowledgements
+        """
+        Flag to disable the requirement that the Executor responds to ACK
+        messages
+        """
+
+        self.framework_acks_sent = 0
+        """
+        Running total of how many acks have been sent to the executor
+        """
+
+        self.framework_acks_received = 0
+        """
+        Running total of how many acks have been received by the executor
+        """
+
+        self.tests = []
+        """
+        A collection of TestCases this scheduler will run through should deploy.
+        """
+
+        self.offers = set()
+        """
+        Keep track of offers available to this framework. Offers are unique by
+        their slave-id.
+        """
+
+    @property
+    def all_tasks(self):
+        """
+        Get all tasks across all tests.
+        """
+        tasks = []
+        for test in self.tests:
+            tasks.extend(test.tasks)
+        return tasks
+
+    @property
+    def all_tests_complete(self):
+        return len([test for test in self.tests if test.complete]) == len(
+            self.tests)
+
+    def kill_test(self, test, msg=None):
+        # Mark the test as done, and explain why
+        test.killed = True
+        test.additional_info = msg
+        if msg:
+            _log.error("Test Killed: %s. Reason: %s", test, msg)
+
+        # Unreserve any offers reserved by this test
+        for offer in self.offers:
+            if offer.reserved_by == test:
+                offer.reserved_by = None
+
+        # If there's no other tests, we're done here
+        if self.all_tests_complete:
+            self.report_results_and_exit()
+
+    def registered(self, driver, frameworkId, masterInfo):
+        """
+        Callback used when the framework is succesfully registered.
+        """
+        _log.info("REGISTERED: with framework ID %s", frameworkId.value)
+
+    def resourceOffers(self, driver, offers):
+        """
+        Triggered when the framework is offered resources by mesos.
+        """
+        for offer in offers:
+            new_offer = Offer(offer)
+            if new_offer.num_tasks_launchable == 0:
+                _log.info("REJECTING EMPTY OFFER: %s", new_offer)
+                driver.declineOffer(offer.id)
+            else:
+                _log.info("New: %s", new_offer)
+                self.offers.add(new_offer)
+        for offer in self.offers:
+            _log.debug(offer)
+
+        # First, hand all offers to the running tests
+        _log.info("Offering to Running Tests")
+        for test in [test for test in self.tests if test.running]:
```

Comment:
If, say, a test reserves a single offer to launch a PingTask and SleepTask, it will accept the Offer with only the SleepTask. Then once the SleepTask is running, and the Slave has reoffered the remaining resources, it will launch the PingTask.

That's where the uncertainty comes in. The Framework assumes that if a Slave offers enough resources to launch 2 apps and it launches just 1, then another offer will come in from the same slave to launch 1 more app after.